### PR TITLE
Branches/concurrency

### PR DESCRIPTION
**Pull request by cmikk**
Run up to $SPOOLDO_JOBS processes in parallel while processing the spool.

### CODE REVIEW

File: spooldo [hunk 1]
```diff
@@ -59,23 +71,35 @@ def archive_incoming(incoming_fname, fname, archive_dnames=[]):
...
-            warn('unable to link %s to %s' % (incoming_fname, active_fname))
-
-        if cmd_tmpl:
-            rc = run_cmd(CommandTemplate(cmd_tmpl).substitute(fname=active_fname))
-            if rc != 0:
-                os.unlink(active_fname)
-                error('command failed')
+    jobs = dict()
+    try:
+        limit = int(os.getenv("SPOOLDO_JOBS"))
```
Comment (djw1149) [inline spooldo]: I would suggest a comment that SPOOLDO_JOBS is optional.
Handle cases of limit < 1 or too high

File: README [hunk 2]
```diff
@@ -18,5 +18,8 @@ be used to construct a command that will be executed for each file as it
 passes through the "active" directory. The sequence "%fname" in the command
 template will be converted to the path to the file to be processed.
 
+If the SPOOLDO_JOBS environment variable is set to a positive integer,
+spooldo will run up to the given number of commands concurrently.
+
 If the SPOOLDO_ORDERED environment variable is set, files in the incoming
```
Comment (djw1149) [inline README:24]: If both _JOBS and _ORDERED are set?  Presumably jobs will be *started* in sorted order?
What is the sort key anyway?

File: spooldo [hunk 3]
```diff
@@ -24,9 +25,20 @@ import time
...
-    return os.system(cmd)
+    return subprocess.Popen(cmd, shell=True).pid
+
+def waitproc(jobs):
+    (pid, status) = os.wait()
+    if pid in jobs:
+        incoming_fname, active_fname, fname, archive_dnames = jobs[pid]
+        if status != 0:
+            error('command failed')
+            os.unlink(active_fname)
```
Comment (djw1149) [inline spooldo:38]: Do you need a try / catch around the unlink

File: spooldo [hunk 4]
```diff
@@ -47,33 +59,47 @@ def same_filesystem(fname1, fname2):
...
 
-        if cmd_tmpl:
-            rc = run_cmd(CommandTemplate(cmd_tmpl).substitute(fname=active_fname))
-            if rc != 0:
-                os.unlink(active_fname)
-                error('command failed')
+def do_spool(incoming_dname, active_dname, archive_dnames=[], cmd_tmpl=None):
+    jobs = dict()
+    try:
+        limit = int(os.getenv("SPOOLDO_JOBS"))
```
Comment (djw1149) [inline spooldo]: Add comment that SPOOLDO_JOBS is optional

File: README (hunk 2 above)
Comment (cmikk) [reply README:24]: Yes, the ordering would apply to the job start. Ordering the job *finishing* is left to the command.

The sort is a lexicographical sort on the filename. This hasn't changed, so I left the documentation untouched for now.

File: spooldo (hunk 1 above)
Comment (cmikk) [reply spooldo]: Will restrict the value to be >=1.

"Too high" will be system dependent, and should not be hard coded.

README introduces SPOOLDO_JOBS with "if SPOOLDO_JOBS is set ..." which shows optional, as does the "try / set default on exception" stanza.

**Review by djw1149 — APPROVED** [review]

**Pull request merged**